### PR TITLE
Initial implementation of codespaces API

### PR DESCRIPTION
**Pull request by artificial-aidan**
https://github.com/google/go-github/issues/2362

### CODE REVIEW

File: github/codespaces.go
Comment (gmlewis) [inline github/codespaces.go:237]:
```suggestion
// Delete deletes a user's codespace.
//
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Nice job, @artificial-aidan!
Just a few changes please, and then we will be ready for a second LGTM+Approval before merging.

**Comment by artificial-aidan** [comment]
> Nice job, @artificial-aidan! Just a few changes please, and then we will be ready for a second LGTM+Approval before merging.

I'll push those up. I also renamed some of the list methods to match others in the repo.

**Comment by artificial-aidan** [comment]
> Thank you! Please run "go generate ./..." in the top of the repo and push (not force-push) the changes to the PR. (...and also run gofmt, of course.) Also, please remember to sign the CLA. See CONTRIBUTING.md for more details.

Working on the CLA now.

Should I push the suggested changes, then run go generate?

File: github/codespaces.go
Comment (artificial-aidan) [reply github/codespaces.go:73]: It's in the response schema for the codespace object: https://docs.github.com/en/rest/codespaces/codespaces?apiVersion=2022-11-28#list-codespaces-in-a-repository-for-the-authenticated-user

image

File: github/codespaces.go
Comment (artificial-aidan) [reply github/codespaces.go:149]: Thanks. I lost this docstring in a refactor it looks like.

**Comment by artificial-aidan** [comment]
The last commit basically mirrors the functionality and layout of Action secrets, but consolidated the tests a bit. I added some examples I used when validating its functionality against a real github.

**Comment by gmlewis** [comment]
Please don't use force-push in this repo, as we always squash-and-merge at the end anyway.
It makes reviews on GitHub particularly painful to see what changed since the last review.
See CONTRIBUTING.md for more details.

**Comment by artificial-aidan** [comment]
> Please don't use force-push in this repo, as we always squash-and-merge at the end anyway. It makes reviews on GitHub particularly painful to see what changed since the last review. See CONTRIBUTING.md for more details.

Sorry about that. Must have been out of habit 👍

File: example/codespaces/newreposecretwithxcrypto/main.go
Comment (gmlewis) [inline example/codespaces/newreposecretwithxcrypto/main.go:118]:
```suggestion
// Sixth, The other two properties of the github.EncodedSecret type are determined. The name of the secret to be added
```

File: example/codespaces/newreposecretwithxcrypto/main.go
Comment (gmlewis) [inline example/codespaces/newreposecretwithxcrypto/main.go:102]: Finally?

File: example/codespaces/newusersecretwithxcrypto/main.go
Comment (gmlewis) [inline example/codespaces/newusersecretwithxcrypto/main.go:111]:
```suggestion
// Sixth, The other two properties of the github.EncodedSecret type are determined. The name of the secret to be added
```

File: github/codespaces.go
Comment (gmlewis) [reply github/codespaces.go:152]: So I guess I wasn't clear.
Please add back "Geo" (with the comments I added) and remove "Location" since "Location" is being deprecated and is getting replaced by "Geo" according to the official docs.

**Review by gmlewis — CHANGES_REQUESTED** [review]

File: github/codespaces.go
Comment (artificial-aidan) [reply github/codespaces.go:152]: That makes a log more sense 😆

Comment (gmlewis) [reply github/codespaces.go:152]: Ah, I see that my "places it" => "replaces it" typo didn't help with my clarity.  :joy:  
Sorry about that!

File: example/codespaces/newusersecretwithxcrypto/main.go
Comment (artificial-aidan) [reply example/codespaces/newusersecretwithxcrypto/main.go:111]: Thanks. Copy paste error from the actions secrets example.

**Review by gmlewis — APPROVED** [review]
Thank you, @artificial-aidan !
LGTM.

Awaiting CLA and LGTM+Approval from any other contributor to this repo before merging.

**Comment by gmlewis** [comment]
It looks like you might need to run `gofmt` and `go generate ./...` again.

**Comment by artificial-aidan** [comment]
@gmlewis got the CLA signed

**Review by gmlewis — APPROVED** [review]
Thank you, @artificial-aidan !
LGTM.

Awaiting LGTM+Approval from any other contributor to this repo before merging.

**Review by valbeat — APPROVED** [review]

**Comment by gmlewis** [comment]
Thank you, @valbeat !
Merging.

**Pull request merged**
